### PR TITLE
Add `netmask()/broadcast()/network()` method on IPv4Cidr #144

### PR DESCRIPTION
**Pull request by LuoZijun**
Add method:

*    netmask()
*    broadcast()
*    network()

### CODE REVIEW

File: src/wire/ipv4.rs [hunk 1]
```diff
@@ -120,6 +120,45 @@ impl Cidr {
...
+                addr[idx] = self.address.0[idx];
+            } else {
+                addr[idx] = 255u8;
+            }
+        }
+        Address(addr)
+    }
+
+    /// Return the network CIDR of this IPv4 CIDR block.
+    pub fn network(&self) -> Cidr {
```
Comment (whitequark) [inline src/wire/ipv4.rs]: This method doesn't make any sense to me. What is "network CIDR"? `self` is already a CIDR.

File: src/wire/ipv4.rs [hunk 2]
```diff
@@ -120,6 +120,45 @@ impl Cidr {
         self.prefix_len
     }
 
+    /// Return the netmask address of this IPv4 CIDR block.
```
Comment (whitequark) [inline src/wire/ipv4.rs]: It's not "netmask address", it's "network mask".

File: src/wire/ipv4.rs [hunk 3]
```diff
@@ -120,6 +120,45 @@ impl Cidr {
...
+    pub fn netmask(&self) -> Address {
+        let mut addr = [0u8; 4];
+        if self.prefix_len >= 8 {
+            for idx in 0usize..((self.prefix_len - self.prefix_len % 8) / 8) as usize {
```
Comment (whitequark) [inline src/wire/ipv4.rs]: Your computations are incorrect for prefix lengths that are not a multiple of 8. There is no requirement that prefix length be a multiple of 8.

File: src/wire/ipv4.rs [hunk 4]
```diff
@@ -814,5 +853,36 @@ mod test {
 
         let cidr_without_prefix = Cidr::new(cidr.address(), 0);
         assert!(cidr_without_prefix.contains_addr(&Address::new(127, 0, 0, 1)));
+
+        // test_cidr_netmask
```
Comment (whitequark) [inline src/wire/ipv4.rs]: These should be separate methods.

File: src/wire/ipv4.rs (hunk 1 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: CIDR not eq network.
```
cirdr = 192.168.1.100/24
network = 192.168.1.0/24
```

File: src/wire/ipv4.rs (hunk 3 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: just do XOR ?

File: src/wire/ipv4.rs (hunk 1 above)
Comment (whitequark) [reply src/wire/ipv4.rs]: OK, the documentation is confusing. How about:

    /// Return the network block of this IPv4 CIDR.

File: src/wire/ipv4.rs [hunk 5]
```diff
@@ -814,5 +853,36 @@ mod test {
...
+        assert_eq!(Cidr::new(Address([127, 0, 0, 0]), 8).netmask(),
+                   Address([255, 0, 0, 0]));
+        assert_eq!(Cidr::new(Address([192, 168, 0, 0]), 16).netmask(),
+                   Address([255, 255, 0, 0]));
+        assert_eq!(Cidr::new(Address([192, 168, 1, 1]), 16).netmask(),
+                   Address([255, 255, 0, 0]));
+        assert_eq!(Cidr::new(Address([192, 168, 1, 1]), 17).netmask(),
+                   Address([255, 255, 0, 0]));
```
Comment (whitequark) [inline src/wire/ipv4.rs]: This test is incorrect. It should be:

```
assert_eq!(Cidr::new(Address([192, 168, 1, 1]), 17).netmask(),
                     Address([255, 255, 128, 0]));
```

File: src/wire/ipv4.rs (hunk 3 above)
Comment (whitequark) [reply src/wire/ipv4.rs]: I don't see how this helps. I've commented below on a test to show how this method should work.

File: src/wire/ipv4.rs (hunk 1 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: ok

File: src/wire/ipv4.rs (hunk 3 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: now should be correct。

File: src/wire/ipv4.rs (hunk 5 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: now should be correct。

**Comment by LuoZijun** [comment]
does need a method `with_netmask()` ?

```
imp Cidr {
    
    pub fn with_netmask(addr: Address, netmask: Address) -> Cidr {
        let prefix_len = netmask.0[0].count_ones()
                         + netmask.0[1].count_ones() 
                         + netmask.0[2].count_ones()
                         + netmask.0[3].count_ones();
        Cidr { address: addr, prefix_len: prefix_len }
    }
}
```

**Comment by whitequark** [comment]
To be honest, I don't know why you even bother with netmasks in the first place. Netmasks are an outdated concept and they shouldn't really be used today.

**Comment by LuoZijun** [comment]
in real world, when you doing `unix network programming` , the `netmask` always follow you.

that's so sad.

**Comment by LuoZijun** [comment]
so, if you have no objection, I will add ' with_netmask ' method.

that's very useful for system network programming.

File: src/wire/ipv4.rs [hunk 6]
```diff
@@ -120,6 +120,45 @@ impl Cidr {
         self.prefix_len
     }
 
+    /// Return the network mask of this IPv4 CIDR block.
+    pub fn netmask(&self) -> Address {
+        let number = 0xffffffffu32 << (32 - self.prefix_len);
```
[the remark this answers is not in the document]
Comment (LuoZijun) [reply src/wire/ipv4.rs]: ok, i will add more test.

File: src/wire/ipv4.rs [hunk 7]
```diff
@@ -120,6 +120,45 @@ impl Cidr {
...
+        ];
+        Address(data)
+    }
+
+    /// Return the broadcast address of this IPv4 CIDR block.
+    pub fn broadcast(&self) -> Address {
+        let network = self.network();
+        let network_address = network.address.0;
+        let number = NetworkEndian::read_u32(&network_address[..]) 
+                    | (0xffffffffu32 << (32 - network.prefix_len) ^ 0xffffffff);
```
[the remark this answers is not in the document]
Comment (LuoZijun) [reply src/wire/ipv4.rs]: 👍

File: src/wire/ipv4.rs [hunk 8]
```diff
@@ -110,6 +110,19 @@ impl Cidr {
...
 
+    /// Create an IPv4 CIDR block from the given address and network mask.
+    ///
+    /// # Panics
+    /// This function panics if the prefix length is larger than 32.
+    pub fn with_netmask(addr: Address, netmask: Address) -> Cidr {
+        let prefix_len = netmask.0[0].count_ones()
+                         + netmask.0[1].count_ones() 
+                         + netmask.0[2].count_ones()
+                         + netmask.0[3].count_ones();
```
Comment (whitequark) [inline src/wire/ipv4.rs]: I think this function must check that the ones in netmask are all consecutive and left-aligned, and return a `Result`.

File: src/wire/ipv4.rs [hunk 9]
```diff
@@ -110,6 +110,19 @@ impl Cidr {
...
+        assert!(prefix_len <= 32);
```
Comment (whitequark) [inline src/wire/ipv4.rs]: This assertion follows trivially from the line above, so it's not necessary.

File: src/wire/ipv4.rs [hunk 10]
```diff
@@ -120,6 +133,53 @@ impl Cidr {
...
+            ((number >>  0) & 0xff) as u8,
+        ];
+        Address(data)
+    }
+
+    /// Return the broadcast address of this IPv4 CIDR.
+    pub fn broadcast(&self) -> Address {
+        let network = self.network();
+
+        if network.prefix_len == 32 {
```
Comment (whitequark) [inline src/wire/ipv4.rs]: Actually, this is not right. A /31 subnet or /32 "subnet" [does not have a broadcast address](http://packetlife.net/blog/2008/jun/18/using-31-bit-subnets-on-point-point-links/). So, this function should return an `Option<Address>`.

File: src/wire/ipv4.rs [hunk 11]
```diff
@@ -110,6 +110,19 @@ impl Cidr {
         Cidr { address, prefix_len }
     }
 
+    /// Create an IPv4 CIDR block from the given address and network mask.
+    ///
+    /// # Panics
+    /// This function panics if the prefix length is larger than 32.
```
Comment (whitequark) [inline src/wire/ipv4.rs]: This function never panics (as written).

File: src/wire/ipv4.rs (hunk 8 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: wow, you are very professional.

File: src/wire/ipv4.rs (hunk 10 above)
Comment (LuoZijun) [reply src/wire/ipv4.rs]: 👍

**Pull request merged**